src/support.cc

```diff
@@ -528,135 +528,264 @@ bool digest_message(const char* alg, const byte* message, int message_len,
-      return false;
-    }
-  } else {
-    printf("authenticated_decrypt: unsupported HMAC\n");
+  HMAC(EVP_sha384(), &key[key_size / 2], mac_size, in, msg_with_iv_size, (byte*)hmac_out, &hmac_size);
+  if (memcmp(hmac_out, in + msg_with_iv_size, mac_size) != 0) {
+    printf("aes_256_cbc_sha384_decrypt: HMAC failed\n");
     return false;
   }
 
   if (!decrypt(in + block_size, msg_with_iv_size - block_size, key, in, out, &plain_size)) {
-    printf("authenticated_decrypt: decrypt failed\n");
+    printf("aes_256_cbc_sha384_decrypt: decrypt failed\n");
     return false;
   }
   *out_size = plain_size;
   return (memcmp(hmac_out, in + msg_with_iv_size, mac_size) == 0);
 }
 
+// We use 128 bit tag
+bool aes_256_gcm_encrypt(byte* in, int in_len, byte *key,
+            byte *iv, byte *out, int* out_size) {
+  EVP_CIPHER_CTX *ctx = nullptr;
+  int len;
+  int ciphertext_len;
+  int blk_size =  cipher_block_byte_size("aes-256");
+  int key_size =  cipher_key_byte_size("aes-256");
+  int tag_len = 0;
+  byte tag[16];
+  int aad_len = 0;
+  byte* aad = nullptr;
+  bool ret = true;
+
+  if(!(ctx = EVP_CIPHER_CTX_new())) {
+    return false;
+  }
+
+  if(1 != EVP_EncryptInit_ex(ctx, EVP_aes_256_gcm(), nullptr, nullptr, nullptr)) {
+    ret = false;
+    goto done;
+  }
+
+  // set IV length
+  if(1 != EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_SET_IVLEN, blk_size, nullptr)) {
+    ret = false;
+    goto done;
+  }
+  if(1 != EVP_EncryptInit_ex(ctx, nullptr, nullptr, key, iv)) {
+    ret = false;
+    goto done;
+  }
+
+  memcpy(out, iv, blk_size);
+
+  if(1 != EVP_EncryptUpdate(ctx, nullptr, &len, aad, aad_len)) {
+    ret = false;
+    goto done;
+  }
+  if(1 != EVP_EncryptUpdate(ctx, out + blk_size, &len, in, in_len)) {
+    ret = false;
+    goto done;
+  }
+  ciphertext_len = len + blk_size;
+
+  // Finalize
+  if(1 != EVP_EncryptFinal_ex(ctx, out + len, &len)) {
+    ret = false;
+    goto done;
+  }
+  ciphertext_len += len;
+
+  tag_len = EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_GET_TAG, blk_size, tag);
+  if(tag_len <= 0) {
+    ret = false;
+    goto done;
+  }
+
+  // append tag
+  memcpy(out + ciphertext_len, tag, blk_size);
+  *out_size = ciphertext_len + blk_size;
+
+done:
+  if (ctx != nullptr) {
+    EVP_CIPHER_CTX_free(ctx);
+    ctx = nullptr;
+  }
+  return ret;
+}
+
+// We use 128 bit tag
+bool aes_256_gcm_decrypt(byte* in, int in_len, byte *key,
+            byte *out, int* out_size) {
+  EVP_CIPHER_CTX *ctx = nullptr;
+  int blk_size =  cipher_block_byte_size("aes-256");
+  int key_size =  cipher_key_byte_size("aes-256");
+  byte* iv = in;
+  bool ret = true;
+  byte* tag = in + in_len - blk_size;
+  int aad_len = 0;
+  byte* aad = nullptr;
+  int len;
+  int plaintext_len;
+  int stream_len = in_len - 2 * blk_size;
+  int err = 0;
+
+  if(!(ctx = EVP_CIPHER_CTX_new())) {
+    return false;
+  }
+  if(!EVP_DecryptInit_ex(ctx, EVP_aes_256_gcm(), nullptr, nullptr, nullptr)) {
+    ret = false;
```

Comment:
Suggestion: In this new chunk of code, there is a repetition of `ret = false;`. Instead, please consider doing this:

- Delete L704 where `ret` is declared and init'ed to `true`. Move that to L716, just before where `ret` is referenced.
- Just before L716, declare & init `bool ret = false;` once.
- Remove all the other initializations of `ret = false;` and just do `goto done`.
- On L752, just before `done: ` label, add `ret = true;`
 
This will clean up the code and get the right behaviour you were intending with all these `goto`s.